kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/timing/Eventually.kt

```diff
@@ -9,63 +14,116 @@ import kotlin.time.TimeSource
+   exceptionClass: KClass<Exception> = Exception::class,
+   predicate: SuspendingPredicate<T> = { true },
+   f: SuspendingProducer<T>
+): T = Eventually(duration, interval, listener, retries, exceptionClass).invoke(predicate, f)
 
 @OptIn(ExperimentalTime::class)
-suspend fun <T, E : Throwable> eventually(
-   duration: Duration,
-   exceptionClass: KClass<E>,
-   f: suspend () -> T
-): T = eventually(duration, 25.milliseconds, exceptionClass, f)
+suspend fun <T> eventually(duration: Duration, interval: Interval, f: SuspendingProducer<T>): T =
+   eventually(duration, interval, f = f)
 
 @OptIn(ExperimentalTime::class)
-suspend fun <T, E : Throwable> eventually(
-   duration: Duration,
-   poll: Duration,
-   exceptionClass: KClass<E>,
-   f: suspend () -> T
-): T {
-   val end = TimeSource.Monotonic.markNow().plus(duration)
-   var times = 0
-   var firstError: Throwable? = null
-   var lastError: Throwable? = null
-   while (end.hasNotPassedNow()) {
-      try {
-         return f()
-      } catch (e: Throwable) {
-         // we only accept exceptions of type exceptionClass and AssertionError
-         // if we didn't accept AssertionError then a matcher failure would immediately fail this function
-         if (!exceptionClass.isInstance(e) && !AssertionError::class.isInstance(e))
-            throw e
-         if (firstError == null)
-            firstError = e
-         else
-            lastError = e
-      }
-      times++
-      delay(poll.toLongMilliseconds())
-   }
+suspend fun <T> eventually(duration: Duration, predicate: SuspendingPredicate<T>, f: SuspendingProducer<T>): T =
+   eventually(duration, predicate = predicate, f = f)
 
-   val sb = StringBuilder()
-   sb.appendLine("Eventually block failed after ${duration}; attempted $times time(s); $poll delay between attempts")
-   sb.appendLine()
+@OptIn(ExperimentalTime::class)
+suspend fun <T, E : Throwable> eventually(duration: Duration, exceptionClass: KClass<E>, f: SuspendingProducer<T>): T =
+   eventually(duration, exceptionClass = exceptionClass, f = f)
 
-   if (firstError != null) {
-      sb.appendLine("The first error was caused by: ${firstError.message}")
-      sb.appendLine(firstError.stackTraceToString())
-   }
+@OptIn(ExperimentalTime::class)
+suspend fun <T> eventually(duration: Duration, interval: Interval, predicate: SuspendingPredicate<T>, f: SuspendingProducer<T>): T =
+   eventually(duration, interval, predicate = predicate, f = f)
 
-   if (lastError != null) {
-      sb.appendLine("The last error was caused by: ${lastError.message}")
-      sb.appendLine(lastError.stackTraceToString())
-   }
+@OptIn(ExperimentalTime::class)
+suspend fun <T, E : Throwable> eventually(duration: Duration, interval: Interval, exceptionClass: KClass<E>, f: SuspendingProducer<T>): T =
+   eventually(duration, interval, exceptionClass = exceptionClass, f = f)
+
+@OptIn(ExperimentalTime::class)
+@Deprecated("Use eventually with an interval, using Duration based poll is deprecated",
```

Comment:
Can we put an example in the deprecatioin message, and add an extension method to turn a duration into an interval if we don't have one already

Reply:
Ah, the first argument to ReplaceWith is an expression that describes how to replace the call site of the deprecated functions. But I also updated the message to include the fixed, fib, and exp calls to convert a Duration to an Interval 👍